### PR TITLE
ramips: add support for Mercusys MR70X

### PR DESCRIPTION
**Pull request by nicefile**
Add support for Mercusys MR70X

Hardware:
- SoC: MediaTek MT7621DAT (880Mhz)
- RAM: DDR3 128MB (SoC)
- Flash: Winbond W25Q128JV (SPI-NOR 16MB)
- WiFi: MediaTek MT7915 DBDC
- Ethernet: SoC (WAN x1, LAN x3, 1Gb)
- UART: [GND, RX, TX, 3.3V] (115200 8N1, J1)

Installation:
1. Flash factory image. This can be done using stock web ui
2. Connect to OpenWrt with an SSH connection to 192.168.1.1
3. Perform sysupgrade with sysupgrade image

Revert to stock firmware:
- Flash stock firmware via OEM Web UI Recovery mode

Web UI Recovery method:
1. Unplug the router
2. Plug in and hold reset button 5~10 secs
3. Set your computer IP address manually to 192.168.1.x / 255.255.255.0
4. Flash image with web browser to 192.168.1.1

### CODE REVIEW

**Comment by nicefile** [comment]
Since [pull/5079 ](https://github.com/openwrt/openwrt/pull/5079) has no maintainer activity I've took matter into my hands . All suggestions from old PR are added .
@kajoj2 @amzamzamzamz can You please test this PR and comment here ?  If possible also under test kernel 5.15

**Comment by nicefile** [comment]
@kajoj2 File was actually empty due to stupid mistake
TPLINK_BOARD_ID := MR70X  line is now fixed
And it seems that it can be only one ID here . So for now MR70X will take the splendor of being probed.  MR1800X should be supported in the same way like different versions of "tplink-eap2x5" are

**Comment by kajoj2** [comment]
@nicefile I have good news, I compile and flashed image. It works without any issue!

**Comment by qwerin** [comment]
@kajoj2 please upload somewhere binary i make some tests. thx

**Comment by kajoj2** [comment]
[openwrt-ramips-mt7621-mercusys_mr70x.zip](https://github.com/openwrt/openwrt/files/10889347/openwrt-ramips-mt7621-mercusys_mr70x.zip)
Sure, here is binary with luci, luci dashboard and I think batman-adv. I don't have any issues. If you want some more pacages to test, please let me know.

**Review by kajoj2 — APPROVED** [review]
I compiled and flashed, work without any issue

**Comment by Lyceris-chan** [comment]
does this pin layout match the one you guys have? 

https://imgur.com/a/sObTiEJ

looking to obtain serial over uart to recover and diagnose my current issue which has been discussed on the forum here:

https://forum.openwrt.org/t/flashing-mercusys-mr70x/154052

**Comment by nicefile** [comment]
@kajoj2 Can you please share mtd0 bin and mtd3 bin (bootloader and factory) with user @Lyceris-chan from your working MR70X. His device refuse to boot from flash for unknown reason.

**Comment by kajoj2** [comment]
[mtd0.zip](https://github.com/openwrt/openwrt/files/10950494/mtd0.zip)
Sure, here is mdt0. I'm little worry about mtd3 because it contains my personal information.

**Comment by Lyceris-chan** [comment]
hey this issue has already been solved, paul edited my factory partition which fixed the issue more details can be found on the forum thread mentioned earlier

**Comment by Lyceris-chan** [comment]
can confirm it works fine :)

File: target/linux/ramips/dts/mt7621_mercusys_mr70x.dts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	chosen {
+		bootargs = "console=ttyS0,115200";
+		bootargs-override = "console=ttyS0,115200";
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_orange: power_orange {
+			label = "green:power";
```
Comment (musashino205) [inline target/linux/ramips/dts/mt7621_mercusys_mr70x.dts:30]: ```orange:power``` ?

File: target/linux/ramips/dts/mt7621_mercusys_mr70x.dts [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+				read-only;
+			};
+
+		partition@40000 {
+				compatible = "denx,uimage";
+				label = "firmware";
+				reg = <0x40000 0xf60000>;
+			};
+
+		factory: partition@fa0000 {
```
Comment (musashino205) [inline target/linux/ramips/dts/mt7621_mercusys_mr70x.dts:77]: wong indentation on ```partition@...``` lines

File: target/linux/ramips/image/mt7621.mk [hunk 3]
```diff
@@ -1963,18 +1975,6 @@ define Device/snr_snr-cpe-me2-lite
...
-  $(Device/uimage-lzma-loader)
-  IMAGE_SIZE := 16064k
-  DEVICE_VENDOR := SNR
-  DEVICE_MODEL := SNR-CPE-ME2-SFP
-  UIMAGE_NAME := $$(DEVICE_MODEL)
-  DEVICE_PACKAGES := kmod-mt7615-firmware kmod-usb3 \
-	    kmod-sfp kmod-usb-ledtrig-usbport uboot-envtools
-endef
-TARGET_DEVICES += snr_snr-cpe-me2-sfp
-
```
Comment (musashino205) [inline target/linux/ramips/image/mt7621.mk:1977]: why did you delete?

File: target/linux/ramips/image/mt7621.mk [hunk 4]
```diff
@@ -2038,18 +2038,6 @@ define Device/totolink_x5000r
...
-  $(Device/tplink-safeloader)
-  DEVICE_MODEL := Archer AX23
-  DEVICE_VARIANT := v1
-  DEVICE_PACKAGES := kmod-mt7915-firmware -uboot-envtools
-  TPLINK_BOARD_ID := ARCHER-AX23-V1
-  KERNEL := $(KERNEL_DTB) | uImage lzma
-  IMAGE_SIZE := 15744k
-endef
-TARGET_DEVICES += tplink_archer-ax23-v1
-
```
Comment (musashino205) [inline target/linux/ramips/image/mt7621.mk:2052]: ditto

**Comment by musashino205** [comment]
> wong indentation on partition@... lines

This issue still exists. Those nodes are child node of ```partitions``` node, so please insert a tab.

File: target/linux/ramips/dts/mt7621_mercusys_mr70x.dts [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+				label = "factory";
+				reg = <0xfa0000 0x60000>;
+				read-only;
+				
+				compatible = "nvmem-cells";
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				macaddr_factory_8: macaddr@8 {
+				reg = <0x8 0x6>;
```
Comment (musashino205) [inline target/linux/ramips/dts/mt7621_mercusys_mr70x.dts:87]: wong indentation

Comment (nicefile) [reply target/linux/ramips/dts/mt7621_mercusys_mr70x.dts:87]: tab added to line 87

**Comment by adonespitogo** [comment]
Just putting it here to say that I'm excited for this :D Looking forward to this being merged <3

**Comment by DavidB137** [comment]
I ported all changes introduced by this commit to v22.03.3, but I'm not able to compile the firmware due to error mentioned by @kajoj2:

> Hello, I'm tired to build, but I'm getting this error:
> 
> ```
> WARNING: Image file /home/karol/code/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/tmp/openwrt-ramips-mt7621-mercusys_mr70x-squashfs-sysupgrade.bin is too big:  > 16121856
> ```
> 
> I also attach error log [log.txt](https://github.com/openwrt/openwrt/files/10865073/log.txt)
> 
> I'm not sure how to fix it. But if you give me a hint, I am happy to help.

Is any modification of `tplink-safeloader.c` needed?

The only thing I had to change in v22.03.3 is to replace `kmod-mt7915-firmware` with `kmod-mt7915e` package. Build works without any problems except for error above.

**Comment by DavidB137** [comment]
Oh, there's more recent version of `firmware-utils` (`tplink-safeloader.c`) which includes support for MR70X introduced by this commit: https://github.com/openwrt/firmware-utils/commit/affc38e89b3544edc97d05367f8aeafa4c4e73bd

I'll try to patch the [Makefile](https://github.com/openwrt/openwrt/blob/v22.03.3/tools/firmware-utils/Makefile). Hopefully, this will solve it.

**Comment by Lyceris-chan** [comment]
> Oh, there's more recent version of `firmware-utils` (`tplink-safeloader.c`) which includes support for MR70X introduced by this commit: [openwrt/firmware-utils@affc38e](https://github.com/openwrt/firmware-utils/commit/affc38e89b3544edc97d05367f8aeafa4c4e73bd)
> 
> I'll try to patch the [Makefile](https://github.com/openwrt/openwrt/blob/v22.03.3/tools/firmware-utils/Makefile). Hopefully, this will solve it.

I've setup a github workflow to build for the mercusys mr70x over here: https://github.com/Lyceris-chan/openwrt if you don't want to bother building it yourself or having to build locally

**Comment by blocktrron** [comment]
Merged to my staging tree, thanks!

**Comment by DrVektor** [comment]
Title: Help needed: MR70X v1 hard brick - need full flash backup
Hello everyone,
I have a Mercusys MR70X v1 that I was using with OpenWrt. Unfortunately, the ethernet ports stopped working. While trying to fix it via UART, I followed a wrong command suggested by an AI assistant, and that ended up corrupting my U-Boot. Now the device is completely dead (hard brick).

I have a CH341A programmer with a SOP8 clip and I want to bring it back to life by flashing a full 16MB SPI flash image directly to the Winbond W25Q128JV chip.

Would anyone here be kind enough to share a full flash backup (mtd0 + mtd1 + mtd2 + mtd3) from their working MR70X v1 device? I understand that the ART/mtd3 partition contains MAC addresses and calibration data, but I can modify that later. I just need the device to boot again.

I would greatly appreciate any help. Thanks in advance!

Best regards

**Comment by qwerin** [comment]
> Title: Help needed: MR70X v1 hard brick - need full flash backup Hello everyone, I have a Mercusys MR70X v1 that I was using with OpenWrt. Unfortunately, the ethernet ports stopped working. While trying to fix it via UART, I followed a wrong command suggested by an AI assistant, and that ended up corrupting my U-Boot. Now the device is completely dead (hard brick).
> 
> I have a CH341A programmer with a SOP8 clip and I want to bring it back to life by flashing a full 16MB SPI flash image directly to the Winbond W25Q128JV chip.
> 
> Would anyone here be kind enough to share a full flash backup (mtd0 + mtd1 + mtd2 + mtd3) from their working MR70X v1 device? I understand that the ART/mtd3 partition contains MAC addresses and calibration data, but I can modify that later. I just need the device to boot again.
> 
> I would greatly appreciate any help. Thanks in advance!
> 
> Best regards

https://drive.google.com/file/d/1prI56azxPL1BMQ9C7tLtXJgzr6WSCm82/view?usp=sharing
